// @bryntum/calendar-angular/esm2015/lib/bryntum-fullscreen-button.component.js
/**
 * Angular wrapper for Bryntum Fullscreen button
 */
import { Component } from '@angular/core';
import { Button, Fullscreen } from '@bryntum/calendar';
import * as i0 from "@angular/core";
export class BryntumFullscreenButtonComponent {
    constructor(element) {
        this.elementRef = element;
    }
    /**
     * Initializes component
     */
    ngOnInit() {
        if (!Fullscreen.enabled) {
            return;
        }
        this.button = new Button({
            appendTo: this.elementRef.nativeElement,
            icon: 'b-icon b-icon-fullscreen',
            tooltip: 'Fullscreen',
            toggleable: true,
            onToggle: ({ pressed }) => {
                pressed ? Fullscreen.request(document.documentElement) : Fullscreen.exit();
            }
        });
        Fullscreen.onFullscreenChange(() => {
            this.button['pressed'] = Fullscreen.isFullscreen;
        });
    }
    /**
     * Destroys component
     */
    ngOnDestroy() {
        if (this.button) {
            this.button.destroy();
        }
    }
}
BryntumFullscreenButtonComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumFullscreenButtonComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
BryntumFullscreenButtonComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.0", type: BryntumFullscreenButtonComponent, selector: "bryntum-fullscreen-button", ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumFullscreenButtonComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'bryntum-fullscreen-button',
                    template: ''
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; } });
