// @bryntum/calendar-angular/esm2015/lib/bryntum-demo-header.component.js
/**
 * Angular wrapper for Bryntum Demo Header
 */
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./bryntum-button.component";
import * as i2 from "./bryntum-fullscreen-button.component";
export class BryntumDemoHeaderComponent {
    constructor(element) {
        this.title = document.title.split(' - ')[1] || document.title;
        this.link = '';
        this.elementRef = element;
        const match = /(.*?\/)(frameworks\/.*?)\/dist/.exec(document.location.href);
        this.link = match ? `${match[1]}#example-${match[2].replace(/\//gm, '-')}` : '#';
    }
}
BryntumDemoHeaderComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumDemoHeaderComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
BryntumDemoHeaderComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.0", type: BryntumDemoHeaderComponent, selector: "bryntum-demo-header", ngImport: i0, template: `
        <header class="demo-header">
            <div id="title-container">
                <a id="title" href="{{ link }}">
                    <h1>{{ title }}</h1>
                </a>
            </div>
            <ng-content></ng-content>
            <div id="tools">
                <bryntum-button
                    text="Download Trial"
                    href="https://bryntum.com/download/?product=calendar"
                    cls="b-green b-raised"
                ></bryntum-button>
                <bryntum-fullscreen-button></bryntum-fullscreen-button>
            </div>
        </header>
`, isInline: true, components: [{ type: i1.BryntumButtonComponent, selector: "bryntum-button", inputs: ["adopt", "align", "anchor", "ariaDescription", "ariaLabel", "bubbleEvents", "callOnFunctions", "centered", "config", "constrainTo", "contentElementCls", "defaultBindProperty", "dock", "draggable", "floating", "hideAnimation", "htmlCls", "ignoreParentReadOnly", "listeners", "localeClass", "localizable", "localizableProperties", "maskDefaults", "masked", "monitorResize", "owner", "positioned", "preventTooltipOnTouch", "relayStoreEvents", "ripple", "rootElement", "scrollAction", "showAnimation", "showTooltipWhenDisabled", "tab", "tag", "textAlign", "title", "ui", "weight", "alignSelf", "appendTo", "badge", "cls", "color", "content", "dataset", "disabled", "extraData", "flex", "height", "hidden", "href", "html", "icon", "iconAlign", "id", "insertBefore", "insertFirst", "margin", "maxHeight", "maxWidth", "menu", "menuIcon", "minHeight", "minWidth", "pressed", "pressedIcon", "readOnly", "scrollable", "target", "text", "toggleable", "toggleGroup", "tooltip", "width", "x", "y", "anchorSize", "type"], outputs: ["onAction", "onBeforeDestroy", "onBeforeHide", "onBeforeShow", "onBeforeShowMenu", "onBeforeToggle", "onCatchAll", "onClick", "onDestroy", "onFocusIn", "onFocusOut", "onHide", "onPaint", "onReadOnly", "onRecompose", "onResize", "onShow", "onToggle"] }, { type: i2.BryntumFullscreenButtonComponent, selector: "bryntum-fullscreen-button" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumDemoHeaderComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'bryntum-demo-header',
                    template: `
        <header class="demo-header">
            <div id="title-container">
                <a id="title" href="{{ link }}">
                    <h1>{{ title }}</h1>
                </a>
            </div>
            <ng-content></ng-content>
            <div id="tools">
                <bryntum-button
                    text="Download Trial"
                    href="https://bryntum.com/download/?product=calendar"
                    cls="b-green b-raised"
                ></bryntum-button>
                <bryntum-fullscreen-button></bryntum-fullscreen-button>
            </div>
        </header>
`
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; } });
