// @bryntum/calendar-angular/esm2015/lib/calendar.module.js
/**
* Angular Bryntum Calendar Shared module
*/
import { NgModule } from '@angular/core';
import { BryntumFullscreenButtonComponent } from './bryntum-fullscreen-button.component';
import { BryntumDemoHeaderComponent } from './bryntum-demo-header.component';
import { BryntumButtonComponent } from './bryntum-button.component';
import { BryntumButtonGroupComponent } from './bryntum-button-group.component';
import { BryntumCalendarComponent } from './bryntum-calendar.component';
import { BryntumCalendarDatePickerComponent } from './bryntum-calendar-date-picker.component';
import { BryntumCheckboxComponent } from './bryntum-checkbox.component';
import { BryntumChipViewComponent } from './bryntum-chip-view.component';
import { BryntumComboComponent } from './bryntum-combo.component';
import { BryntumContainerComponent } from './bryntum-container.component';
import { BryntumDateFieldComponent } from './bryntum-date-field.component';
import { BryntumDatePickerComponent } from './bryntum-date-picker.component';
import { BryntumDateTimeFieldComponent } from './bryntum-date-time-field.component';
import { BryntumDisplayFieldComponent } from './bryntum-display-field.component';
import { BryntumDurationFieldComponent } from './bryntum-duration-field.component';
import { BryntumFieldFilterPickerComponent } from './bryntum-field-filter-picker.component';
import { BryntumFieldFilterPickerGroupComponent } from './bryntum-field-filter-picker-group.component';
import { BryntumFileFieldComponent } from './bryntum-file-field.component';
import { BryntumFilePickerComponent } from './bryntum-file-picker.component';
import { BryntumFilterFieldComponent } from './bryntum-filter-field.component';
import { BryntumGridComponent } from './bryntum-grid.component';
import { BryntumGridBaseComponent } from './bryntum-grid-base.component';
import { BryntumGridFieldFilterPickerComponent } from './bryntum-grid-field-filter-picker.component';
import { BryntumGridFieldFilterPickerGroupComponent } from './bryntum-grid-field-filter-picker-group.component';
import { BryntumListComponent } from './bryntum-list.component';
import { BryntumMenuComponent } from './bryntum-menu.component';
import { BryntumModeSelectorComponent } from './bryntum-mode-selector.component';
import { BryntumNumberFieldComponent } from './bryntum-number-field.component';
import { BryntumPagingToolbarComponent } from './bryntum-paging-toolbar.component';
import { BryntumPanelComponent } from './bryntum-panel.component';
import { BryntumPasswordFieldComponent } from './bryntum-password-field.component';
import { BryntumProjectComboComponent } from './bryntum-project-combo.component';
import { BryntumProjectModelComponent } from './bryntum-project-model.component';
import { BryntumRadioComponent } from './bryntum-radio.component';
import { BryntumRadioGroupComponent } from './bryntum-radio-group.component';
import { BryntumResourceComboComponent } from './bryntum-resource-combo.component';
import { BryntumResourceFilterComponent } from './bryntum-resource-filter.component';
import { BryntumResourceViewComponent } from './bryntum-resource-view.component';
import { BryntumSchedulerDatePickerComponent } from './bryntum-scheduler-date-picker.component';
import { BryntumSidebarComponent } from './bryntum-sidebar.component';
import { BryntumSliderComponent } from './bryntum-slider.component';
import { BryntumSlideToggleComponent } from './bryntum-slide-toggle.component';
import { BryntumSplitterComponent } from './bryntum-splitter.component';
import { BryntumTabPanelComponent } from './bryntum-tab-panel.component';
import { BryntumTextAreaFieldComponent } from './bryntum-text-area-field.component';
import { BryntumTextAreaPickerFieldComponent } from './bryntum-text-area-picker-field.component';
import { BryntumTextFieldComponent } from './bryntum-text-field.component';
import { BryntumTimeFieldComponent } from './bryntum-time-field.component';
import { BryntumTimePickerComponent } from './bryntum-time-picker.component';
import { BryntumToolbarComponent } from './bryntum-toolbar.component';
import { BryntumTreeComboComponent } from './bryntum-tree-combo.component';
import { BryntumUndoRedoComponent } from './bryntum-undo-redo.component';
import { BryntumViewPresetComboComponent } from './bryntum-view-preset-combo.component';
import { BryntumWidgetComponent } from './bryntum-widget.component';
import { BryntumYearPickerComponent } from './bryntum-year-picker.component';
import * as i0 from "@angular/core";
export class BryntumCalendarModule {
}
BryntumCalendarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumCalendarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BryntumCalendarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumCalendarModule, declarations: [BryntumFullscreenButtonComponent,
        BryntumDemoHeaderComponent,
        BryntumButtonComponent,
        BryntumButtonGroupComponent,
        BryntumCalendarComponent,
        BryntumCalendarDatePickerComponent,
        BryntumCheckboxComponent,
        BryntumChipViewComponent,
        BryntumComboComponent,
        BryntumContainerComponent,
        BryntumDateFieldComponent,
        BryntumDatePickerComponent,
        BryntumDateTimeFieldComponent,
        BryntumDisplayFieldComponent,
        BryntumDurationFieldComponent,
        BryntumFieldFilterPickerComponent,
        BryntumFieldFilterPickerGroupComponent,
        BryntumFileFieldComponent,
        BryntumFilePickerComponent,
        BryntumFilterFieldComponent,
        BryntumGridComponent,
        BryntumGridBaseComponent,
        BryntumGridFieldFilterPickerComponent,
        BryntumGridFieldFilterPickerGroupComponent,
        BryntumListComponent,
        BryntumMenuComponent,
        BryntumModeSelectorComponent,
        BryntumNumberFieldComponent,
        BryntumPagingToolbarComponent,
        BryntumPanelComponent,
        BryntumPasswordFieldComponent,
        BryntumProjectComboComponent,
        BryntumProjectModelComponent,
        BryntumRadioComponent,
        BryntumRadioGroupComponent,
        BryntumResourceComboComponent,
        BryntumResourceFilterComponent,
        BryntumResourceViewComponent,
        BryntumSchedulerDatePickerComponent,
        BryntumSidebarComponent,
        BryntumSliderComponent,
        BryntumSlideToggleComponent,
        BryntumSplitterComponent,
        BryntumTabPanelComponent,
        BryntumTextAreaFieldComponent,
        BryntumTextAreaPickerFieldComponent,
        BryntumTextFieldComponent,
        BryntumTimeFieldComponent,
        BryntumTimePickerComponent,
        BryntumToolbarComponent,
        BryntumTreeComboComponent,
        BryntumUndoRedoComponent,
        BryntumViewPresetComboComponent,
        BryntumWidgetComponent,
        BryntumYearPickerComponent], exports: [BryntumFullscreenButtonComponent,
        BryntumDemoHeaderComponent,
        BryntumButtonComponent,
        BryntumButtonGroupComponent,
        BryntumCalendarComponent,
        BryntumCalendarDatePickerComponent,
        BryntumCheckboxComponent,
        BryntumChipViewComponent,
        BryntumComboComponent,
        BryntumContainerComponent,
        BryntumDateFieldComponent,
        BryntumDatePickerComponent,
        BryntumDateTimeFieldComponent,
        BryntumDisplayFieldComponent,
        BryntumDurationFieldComponent,
        BryntumFieldFilterPickerComponent,
        BryntumFieldFilterPickerGroupComponent,
        BryntumFileFieldComponent,
        BryntumFilePickerComponent,
        BryntumFilterFieldComponent,
        BryntumGridComponent,
        BryntumGridBaseComponent,
        BryntumGridFieldFilterPickerComponent,
        BryntumGridFieldFilterPickerGroupComponent,
        BryntumListComponent,
        BryntumMenuComponent,
        BryntumModeSelectorComponent,
        BryntumNumberFieldComponent,
        BryntumPagingToolbarComponent,
        BryntumPanelComponent,
        BryntumPasswordFieldComponent,
        BryntumProjectComboComponent,
        BryntumProjectModelComponent,
        BryntumRadioComponent,
        BryntumRadioGroupComponent,
        BryntumResourceComboComponent,
        BryntumResourceFilterComponent,
        BryntumResourceViewComponent,
        BryntumSchedulerDatePickerComponent,
        BryntumSidebarComponent,
        BryntumSliderComponent,
        BryntumSlideToggleComponent,
        BryntumSplitterComponent,
        BryntumTabPanelComponent,
        BryntumTextAreaFieldComponent,
        BryntumTextAreaPickerFieldComponent,
        BryntumTextFieldComponent,
        BryntumTimeFieldComponent,
        BryntumTimePickerComponent,
        BryntumToolbarComponent,
        BryntumTreeComboComponent,
        BryntumUndoRedoComponent,
        BryntumViewPresetComboComponent,
        BryntumWidgetComponent,
        BryntumYearPickerComponent] });
BryntumCalendarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumCalendarModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.0", ngImport: i0, type: BryntumCalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        BryntumFullscreenButtonComponent,
                        BryntumDemoHeaderComponent,
                        BryntumButtonComponent,
                        BryntumButtonGroupComponent,
                        BryntumCalendarComponent,
                        BryntumCalendarDatePickerComponent,
                        BryntumCheckboxComponent,
                        BryntumChipViewComponent,
                        BryntumComboComponent,
                        BryntumContainerComponent,
                        BryntumDateFieldComponent,
                        BryntumDatePickerComponent,
                        BryntumDateTimeFieldComponent,
                        BryntumDisplayFieldComponent,
                        BryntumDurationFieldComponent,
                        BryntumFieldFilterPickerComponent,
                        BryntumFieldFilterPickerGroupComponent,
                        BryntumFileFieldComponent,
                        BryntumFilePickerComponent,
                        BryntumFilterFieldComponent,
                        BryntumGridComponent,
                        BryntumGridBaseComponent,
                        BryntumGridFieldFilterPickerComponent,
                        BryntumGridFieldFilterPickerGroupComponent,
                        BryntumListComponent,
                        BryntumMenuComponent,
                        BryntumModeSelectorComponent,
                        BryntumNumberFieldComponent,
                        BryntumPagingToolbarComponent,
                        BryntumPanelComponent,
                        BryntumPasswordFieldComponent,
                        BryntumProjectComboComponent,
                        BryntumProjectModelComponent,
                        BryntumRadioComponent,
                        BryntumRadioGroupComponent,
                        BryntumResourceComboComponent,
                        BryntumResourceFilterComponent,
                        BryntumResourceViewComponent,
                        BryntumSchedulerDatePickerComponent,
                        BryntumSidebarComponent,
                        BryntumSliderComponent,
                        BryntumSlideToggleComponent,
                        BryntumSplitterComponent,
                        BryntumTabPanelComponent,
                        BryntumTextAreaFieldComponent,
                        BryntumTextAreaPickerFieldComponent,
                        BryntumTextFieldComponent,
                        BryntumTimeFieldComponent,
                        BryntumTimePickerComponent,
                        BryntumToolbarComponent,
                        BryntumTreeComboComponent,
                        BryntumUndoRedoComponent,
                        BryntumViewPresetComboComponent,
                        BryntumWidgetComponent,
                        BryntumYearPickerComponent
                    ],
                    imports: [],
                    exports: [
                        BryntumFullscreenButtonComponent,
                        BryntumDemoHeaderComponent,
                        BryntumButtonComponent,
                        BryntumButtonGroupComponent,
                        BryntumCalendarComponent,
                        BryntumCalendarDatePickerComponent,
                        BryntumCheckboxComponent,
                        BryntumChipViewComponent,
                        BryntumComboComponent,
                        BryntumContainerComponent,
                        BryntumDateFieldComponent,
                        BryntumDatePickerComponent,
                        BryntumDateTimeFieldComponent,
                        BryntumDisplayFieldComponent,
                        BryntumDurationFieldComponent,
                        BryntumFieldFilterPickerComponent,
                        BryntumFieldFilterPickerGroupComponent,
                        BryntumFileFieldComponent,
                        BryntumFilePickerComponent,
                        BryntumFilterFieldComponent,
                        BryntumGridComponent,
                        BryntumGridBaseComponent,
                        BryntumGridFieldFilterPickerComponent,
                        BryntumGridFieldFilterPickerGroupComponent,
                        BryntumListComponent,
                        BryntumMenuComponent,
                        BryntumModeSelectorComponent,
                        BryntumNumberFieldComponent,
                        BryntumPagingToolbarComponent,
                        BryntumPanelComponent,
                        BryntumPasswordFieldComponent,
                        BryntumProjectComboComponent,
                        BryntumProjectModelComponent,
                        BryntumRadioComponent,
                        BryntumRadioGroupComponent,
                        BryntumResourceComboComponent,
                        BryntumResourceFilterComponent,
                        BryntumResourceViewComponent,
                        BryntumSchedulerDatePickerComponent,
                        BryntumSidebarComponent,
                        BryntumSliderComponent,
                        BryntumSlideToggleComponent,
                        BryntumSplitterComponent,
                        BryntumTabPanelComponent,
                        BryntumTextAreaFieldComponent,
                        BryntumTextAreaPickerFieldComponent,
                        BryntumTextFieldComponent,
                        BryntumTimeFieldComponent,
                        BryntumTimePickerComponent,
                        BryntumToolbarComponent,
                        BryntumTreeComboComponent,
                        BryntumUndoRedoComponent,
                        BryntumViewPresetComboComponent,
                        BryntumWidgetComponent,
                        BryntumYearPickerComponent
                    ]
                }]
        }] });
